### PR TITLE
WT-9144 Update Wiredtiger cursor documentation

### PR DESCRIPTION
**Pull request by clarissecheah**

### CODE REVIEW

File: src/docs/arch-cursor.dox [hunk 1]
```diff
@@ -371,6 +371,15 @@ how many references to closed data handles are freed, we may continue our walk.
 usually strike a good balance between not having a lot of overhead for sweeps, and keeping up with
 the need to free up shared resources.
 
+@section arch_cursor_bound Range bounded cursors
+
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:375]: @ddanderson does this seem like the right level of detail for the architecture guide?

File: src/docs/cursor-ops.dox [hunk 2]
```diff
@@ -161,4 +161,45 @@ that may not be modified or freed by the application.  If a longer scope
 is required, the application must make a copy of the memory before the
 cursor is re-used, closed or reset.
 
+@section cursor_bound Apply bounds to the cursor
+
+Cursors may have bounds applied to them. This sets a start and stop key for
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:166]: I would update this to be a bit clearer about what bounds are and why we need them.\
```
A cursor may have a set of bounds applied to it, this restricts the key range that the cursor operates within. 

Suppose we have a table with keys A to Z. We can restrict the cursor to only operate over keys L->Q. If the cursor walks past key Q, the cursor will enter an un-positioned state having walked off the "end" of the table from it's perspective. Additionally should the cursor perform a search_near it will only search within the keyrange of L->Q. This can provide a performance boost when operating within tables that have a large number of deleted or invisible records.

The cursor can have a lower bound set or an upper bound set, or both.

```

File: src/docs/arch-cursor.dox [hunk 3]
```diff
@@ -371,6 +371,15 @@ how many references to closed data handles are freed, we may continue our walk.
...
+The lower and upper bound in a cursor is represented as a \c WT_ITEM. The implementation applies bounds through the function
+__wt_cursor_bound which takes the cursor and a config string as arguments. The config string specifies whether the operation is clearing
+or setting the bound, and if a "set" is in progress, whether the bound is a lower or upper bound. There are four flags that can get set
+when applying bounds: WT_CURSTD_BOUND_LOWER, WT_CURSTD_BOUND_LOWER_INCLUSIVE, WT_CURSTD_BOUND_UPPER, and WT_CURSTD_BOUND_UPPER_INCLUSIVE.
+When calling __wt_cursor_bound with valid bounds, the respective bound flag gets set and the bound key data saved in the cursor.
+Clearing a cursor's bounds can be done via resetting the cursor or calling bound with the "clear" config which will free both bounds buffers.
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:381]: I realized we should include a bit more detail here about the performance optimization we get from bounded cursors. 

We should mention that the bounds checking ignores whether the value associated with the key is visible to the cursor and can return `WT_NOTFOUND` when it detects a key that is outside the range.

File: src/docs/arch-cursor.dox [hunk 4]
```diff
@@ -371,6 +371,15 @@ how many references to closed data handles are freed, we may continue our walk.
...
+
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:382]: I would add detail here about how we implemented the bounds position logic in `cursor->next/prev` and also in `cursor->search_near`.

File: src/docs/cursor-ops.dox [hunk 5]
```diff
@@ -161,4 +161,45 @@ that may not be modified or freed by the application.  If a longer scope
...
+```
+cursor->set_key(cursor, A)
+cursor->bound(cursor, bound=lower)
+```
+
+Both or either one of the lower and upper bounds can be set, and the unset bound will be assumed to be off the start or
+end of the key range accordingly. When setting the bounds, the relevant bound is set prior with cursor->set_key.
+This key does not have to exist or be visible in order for bounds setting to succeed.
+
+Cursor operations next, prev, search, and search near support bounds. When performing a traversal such as cursor->next
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:178]: We actually also check bounds in cursor `modify` `insert` `update` `remove`

File: src/docs/arch-cursor.dox (hunk 4 above)
Comment (clarissecheah) [reply src/docs/arch-cursor.dox:382]: Great point, thanks! I've added detail on how we position bounded cursors in both cases here: [6b88eb1](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/6b88eb16841dd50de96ea1436b961ec3a0e8644d).

File: src/docs/cursor-ops.dox (hunk 5 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox]: [6b88eb1](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/6b88eb16841dd50de96ea1436b961ec3a0e8644d)

File: src/docs/arch-cursor.dox (hunk 3 above)
Comment (clarissecheah) [reply src/docs/arch-cursor.dox:381]: I have included mention of the performance optimization we get here [6b88eb1](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/6b88eb16841dd50de96ea1436b961ec3a0e8644d). I did not go very far into detail with it, though, so let me know if I should have written more.

File: src/docs/cursor-ops.dox (hunk 2 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox]: [6b88eb1](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/6b88eb16841dd50de96ea1436b961ec3a0e8644d)

File: src/docs/cursor-ops.dox [hunk 6]
```diff
@@ -161,4 +161,52 @@ that may not be modified or freed by the application.  If a longer scope
...
 cursor is re-used, closed or reset.
 
+@section cursor_bound Apply bounds to the cursor
+
+A cursor may have a set of bounds applied to it, this restricts the key range
+that the cursor operates within.
+
+Suppose we have a table with keys A to Z. We can restrict the cursor to only operate over
+keys L->Q. If the cursor walks past key Q, the cursor will enter an un-positioned state having
+walked off the "end" of the table from it's perspective. Additionally should the cursor perform a
```
Comment (ddanderson) [inline src/docs/cursor-ops.dox:171]: `it's` -> `its`

File: src/docs/cursor-ops.dox [hunk 7]
```diff
@@ -161,4 +161,52 @@ that may not be modified or freed by the application.  If a longer scope
...
+search_near it will only search within the key range of L->Q. This can provide a performance boost
+when operating within tables that have a large number of deleted or invisible records. The cursor
+can have a lower bound set or an upper bound set, or both.
+
+```
+cursor->set_key(cursor, A)
+cursor->bound(cursor, bound=lower)
+```
+
+Both or either one of the lower and upper bounds can be set, and the unset bound will be assumed to be off the start or
```
Comment (ddanderson) [inline src/docs/cursor-ops.dox:181]: and the unset bound -> and **any** unset bound

File: src/docs/cursor-ops.dox [hunk 8]
```diff
@@ -161,4 +161,52 @@ that may not be modified or freed by the application.  If a longer scope
...
+cursor->next or cursor->prev, the cursor will only return keys that are within the bounds that have been set. For example, calling cursor->next
+on an unpositioned range bounded cursor will position it on the lower bound rather than the end of the key range.
+
+Example API usage with cursor traversal:
+
+```
+cursor->set_key(cursor, A)
+cursor->bound(cursor, bound=lower,inclusive=false)
+cursor->next(cursor)
+```
```
Comment (ddanderson) [inline src/docs/cursor-ops.dox:195]: (Optional): The best way to have example code is to put it into one of the programs in examples/c and refer to it here.

File: src/docs/spell.ok [hunk 9]
```diff
@@ -187,6 +187,7 @@ autogen
 automake
 basecfg
 basho
+behaviour
```
Comment (ddanderson) [inline src/docs/spell.ok:190]: I think MongoDB and WT use American spelling (behavior and optimized) in public documents.  I would leave out behaviour and optimised from this list and change the text.

**Review by ddanderson — APPROVED** [review]
LGTM - a few mostly small things to fix or consider.

File: src/docs/cursor-ops.dox [hunk 10]
```diff
@@ -161,4 +161,52 @@ that may not be modified or freed by the application.  If a longer scope
...
+
+A cursor may have a set of bounds applied to it, this restricts the key range
+that the cursor operates within.
+
+Suppose we have a table with keys A to Z. We can restrict the cursor to only operate over
+keys L->Q. If the cursor walks past key Q, the cursor will enter an un-positioned state having
+walked off the "end" of the table from it's perspective. Additionally should the cursor perform a
+search_near it will only search within the key range of L->Q. This can provide a performance boost
+when operating within tables that have a large number of deleted or invisible records. The cursor
+can have a lower bound set or an upper bound set, or both.
```
Comment (ddanderson) [inline src/docs/cursor-ops.dox:174]: I'd add a little more on the performance note:  This can provide **an application with** a performance boost
when operating within tables that have a large number of deleted or invisible records **just outside the boundaries**.

Comment (ddanderson) [reply src/docs/cursor-ops.dox:174]: True, the deleted records can exist within the boundaries, but does the app get a performance benefit over regular cursor traversal?  Assume boundaries at 10 and 20 and there are visible values at 9, 13, 15, 21 with deletions at all numbers in between 11,12,14,16,17,18,19.  Isn't the performance the same as not using boundaries?  I'm sure I'm missing something - if there's something more subtle maybe that's something that could be in the doc.

Comment (luke-pearson) [reply src/docs/cursor-ops.dox:174]: I think you're not missing anything I just think I wanted to be explicit about why you'd get a performance speedup and that's because we check the key before we do any visibility checking. The performance would only optimize cases where the deleted keys are on or after the bounds.

Comment (ddanderson) [reply src/docs/cursor-ops.dox:174]: Yes, makes sense.

Comment (luke-pearson) [reply src/docs/cursor-ops.dox:174]: How about: 
```
This can provide an application with a performance boost when operating within tables that have a large number of deleted or invisible records. This is because the key check occurs before we determine if the value associated with the key is visible or not.
```
That seems like the best of both worlds.

File: src/docs/spell.ok (hunk 9 above)
Comment (clarissecheah) [reply src/docs/spell.ok]: That's true, thanks! [9b21f18](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/9b21f186bb973ba5c89e96925bf8becfecf64b56)

File: src/docs/cursor-ops.dox (hunk 8 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox]: Thanks for the suggestion! [9b21f18](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/9b21f186bb973ba5c89e96925bf8becfecf64b56)

File: src/docs/cursor-ops.dox (hunk 7 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox]: [9b21f18](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/9b21f186bb973ba5c89e96925bf8becfecf64b56)

File: src/docs/cursor-ops.dox (hunk 6 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox]: [9b21f18](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/9b21f186bb973ba5c89e96925bf8becfecf64b56)

File: examples/c/ex_cursor.c [hunk 11]
```diff
@@ -84,6 +85,16 @@ cursor_reset(WT_CURSOR *cursor)
 }
 /*! [cursor reset] */
 
+/*! [cursor bound] */
+int
+cursor_bound(WT_CURSOR *cursor)
+{
+    cursor->set_key(cursor, "A");
+    error_check(cursor->bound(cursor, "bound=lower,inclusive=false"));
```
Comment (luke-pearson) [inline examples/c/ex_cursor.c:93]: I would change this to: "action=set,bound=lower"

File: src/docs/arch-cursor.dox (hunk 3 above)
Comment (luke-pearson) [reply src/docs/arch-cursor.dox:384]: I would rephrase the first few sentences  a bit:
```
Range bounded cursors provide a mechanism for optimizing cursor operations in collections with a large number of invisible records. When traversing records the record's key is compared with the bounds set on the cursor and the traversal will exit with WT_NOTFOUND regardless of whether the value associated with the cursor is visible.
```

File: src/docs/arch-cursor.dox [hunk 12]
```diff
@@ -371,6 +371,28 @@ how many references to closed data handles are freed, we may continue our walk.
...
 the need to free up shared resources.
 
+@section arch_cursor_bound Range bounded cursors
+
+The implementation of bounds brings performance optimizations. Bounds checking detects when a key is out of range and returns WT_NOTFOUND
+regardless of whether the value associated with the key is visible to the cursor. This skips the traversal of unnecessary records.
+
+The lower and upper bound in a cursor is represented as a \c WT_ITEM. The implementation applies bounds through the function
+__wt_cursor_bound which takes the cursor and a config string as arguments. The config string specifies whether the operation is clearing
+or setting the bound, and if a "set" is in progress, whether the bound is a lower or upper bound. There are four flags that can get set
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:381]: Rephrase: "that can get set" -> "that can be set"

File: src/docs/arch-cursor.dox [hunk 13]
```diff
@@ -371,6 +371,28 @@ how many references to closed data handles are freed, we may continue our walk.
...
+when applying bounds: WT_CURSTD_BOUND_LOWER, WT_CURSTD_BOUND_LOWER_INCLUSIVE, WT_CURSTD_BOUND_UPPER, and WT_CURSTD_BOUND_UPPER_INCLUSIVE.
+When calling __wt_cursor_bound with valid bounds, the respective bound flag gets set and the bound key data saved in the cursor.
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:383]: "the respective bound flag gets set and the bound key data saved in the cursor." -> "the respective bound flag is set and the key held on the cursor is copied into the relevant bound buffer."

File: src/docs/arch-cursor.dox [hunk 14]
```diff
@@ -371,6 +371,28 @@ how many references to closed data handles are freed, we may continue our walk.
...
+Clearing a cursor's bounds can be done via resetting the cursor or calling bound with the "clear" config which will free both bounds buffers.
+
+Positioning the cursor for traversal with bounds has a separate implementation to place the cursor at the "start" or "end" of the range it can operate
```
Comment (luke-pearson) [inline src/docs/arch-cursor.dox:386]: Reword first sentence: When calling \c cursor->next and \c cursor->prev on an unpositioned cursor WiredTiger guarantees that the cursor will start walking from the start and end of the file respectively. The range bounded cursor logic continues to uphold this guarantee. Doing so requires additional logic to place the cursor at the start or end of the range.

Comment (clarissecheah) [reply src/docs/arch-cursor.dox]: [f382c49](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/f382c49e92533aafae89dae772c9766b5fbc97ce)

File: src/docs/arch-cursor.dox (hunk 13 above)
Comment (clarissecheah) [reply src/docs/arch-cursor.dox]: [f382c49](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/f382c49e92533aafae89dae772c9766b5fbc97ce)

File: src/docs/arch-cursor.dox (hunk 12 above)
Comment (clarissecheah) [reply src/docs/arch-cursor.dox]: [f382c49](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/f382c49e92533aafae89dae772c9766b5fbc97ce)

File: examples/c/ex_cursor.c (hunk 11 above)
Comment (clarissecheah) [reply examples/c/ex_cursor.c]: [f382c49](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/f382c49e92533aafae89dae772c9766b5fbc97ce)

File: src/docs/cursor-ops.dox (hunk 10 above)
Comment (ddanderson) [reply src/docs/cursor-ops.dox:174]: Great!

File: src/docs/arch-cursor.dox (hunk 1 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:375]: Yes, I think it's at the right level.  When I'm writing/reviewing these sections, I'm thinking about:

- a generic overview of the code (and not a rehash of user documentation), it can be helpful to have the overview before someone dives into one piece of the code.  Often this is a quick discussion of any internal data structure, how it is used, and any "rules" that must be followed when using the data structure.
- somewhat equivalently, those things that an experienced coder might learn from a quick scan of the code.
- any other subtle things that it's hard to see by looking at the code.

Many of the above can/should also appear in code comments.

File: src/docs/cursor-ops.dox [hunk 15]
```diff
@@ -161,4 +161,42 @@ that may not be modified or freed by the application.  If a longer scope
...
+
+Clearing the bounds can be done through resetting the cursor with cursor->reset or by passing a clear config to the bounds API call.
+This will unconditionally clear both bounds.
+
+The bounds API supports these cursor types:
+- File cursor
+- Table cursor
+- Datastore cursor
+- Dump cursor
+- Index cursor
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:199]: Lets remove `index cursor` for now. I also think the JSON cursor is just a subset of the dump cursor so it can also be removed.

We should also mention here that we support row-store and variable length column store but not fixed length column store.

File: src/docs/cursor-ops.dox [hunk 16]
```diff
@@ -161,4 +161,42 @@ that may not be modified or freed by the application.  If a longer scope
...
+keys \c L->Q. If the cursor walks past key \c Q, the cursor will enter an unpositioned state having
+walked off the "end" of the table from its perspective. Additionally should the cursor perform a
+search_near, it will only search within the key range of \c L->Q.  This can provide an application with a
+performance boost when operating within tables that have a large number of deleted or invisible records.
+This is because the key comparison with the bound occurs before we determine if the value associated
+with the key is visible or not. The cursor can have a lower bound set or an upper bound set, or both.
+
+@snippet ex_cursor.c cursor bound
+
+Both or either one of the lower and upper bounds can be set, and any unset bound will be assumed to be off the start or
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:179]: I would rephrase this a bit: Both or either one of the lower and upper bounds can be set, if a bound is not set then the usual start or end of the file applies. Given our previous example had we set a lower bound of \c L and no upper bound then our key range would be \c L->Z.

File: src/docs/cursor-ops.dox [hunk 17]
```diff
@@ -161,4 +161,42 @@ that may not be modified or freed by the application.  If a longer scope
...
+end of the key range accordingly. When setting the bounds, the relevant bound is set prior with \c cursor->set_key.
+This key does not have to exist or be visible in order for bounds setting to succeed.
+
+Cursor operations next, prev, search, search near, modify, insert, update, and remove support bounds. When performing a traversal such as
+\c cursor->next or \c cursor->prev, the cursor will only return keys that are within the bounds that have been set. For example, calling \c cursor->next
+on an unpositioned range bounded cursor will position it on the lower bound rather than the end of the key range.
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:185]: rather than the start of the key range.

File: src/docs/cursor-ops.dox [hunk 18]
```diff
@@ -161,4 +161,42 @@ that may not be modified or freed by the application.  If a longer scope
...
+
+If search is called on a key that's not within bounds, this will result in returning a \c WT_NOTFOUND. \c cursor->search_near behavior is optimized with the
+inclusion of bounds, such that an early exit is made possible once the current key has exceeded the bounds. Calling
```
Comment (luke-pearson) [inline src/docs/cursor-ops.dox:188]: Rephrase: \c cursor->search_near can benefit from range bounded cursors, once the cursor traverses outside of the bounded range the search_near logic can early exit reducing the overall number of keys visited.

**Review by luke-pearson — APPROVED** [review]
LGTM, with one final \c comment.

File: src/docs/cursor-ops.dox [hunk 19]
```diff
@@ -161,4 +161,42 @@ that may not be modified or freed by the application.  If a longer scope
...
+performance boost when operating within tables that have a large number of deleted or invisible records.
+This is because the key comparison with the bound occurs before we determine if the value associated
+with the key is visible or not. The cursor can have a lower bound set or an upper bound set, or both.
+
+@snippet ex_cursor.c cursor bound
+
+Both or either one of the lower and upper bounds can be set, if a bound is not set then the usual start or end of the file applies.
+Given our previous example had we set a lower bound of \c L and no upper bound then our key range would be \c L->Z. When setting the bounds,
+the relevant bound is set prior with \c cursor->set_key. This key does not have to exist or be visible in order for bounds setting to succeed.
+
```
Comment (ddanderson) [inline src/docs/cursor-ops.dox:182]: I think somewhere we should say something about a copy of the key being made when the bounds call is made.  Otherwise, one might think that WT just keeps a pointer to the user's memory (and hence the user needs to keep that memory intact), and that's not the case.  Perhaps this is a good point to add it, something like:
```
When a bounds is set, a copy of the bounds key is made and is kept by the cursor.  Thus the memory used for the key when making the \c cursor->bound call does not need to be retained by the application.
```

**Review by ddanderson — APPROVED** [review]
LGTM.  One suggested addition, that you can take if you like.

File: src/docs/cursor-ops.dox (hunk 19 above)
Comment (clarissecheah) [reply src/docs/cursor-ops.dox:182]: Sounds good, thanks! [77fac8d](https://github.com/wiredtiger/wiredtiger/pull/8367/commits/77fac8d6630882119807efca90bc3dda42d2a1c0)
